src/lib/ams.c

```diff
@@ -0,0 +1,745 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright (c) 2022 Intel Corporation
+//
+// Author: Krzysztof Frydryk <krzysztofx.frydryk@intel.com>
+
+#include <sof/lib/memory.h>
+#include <sof/lib/cpu.h>
+#include <sof/coherent.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/schedule/schedule.h>
+#include <sof/ipc/topology.h>
+#include <sof/drivers/idc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sof/lib/ams.h>
+
+LOG_MODULE_REGISTER(ams, CONFIG_SOF_LOG_LEVEL);
+
+DECLARE_SOF_UUID("ams", ams_uuid, 0xea9c4bca, 0x5b7d, 0x48c6,
+		 0x95, 0x86, 0x55, 0x3e, 0x27, 0x23, 0x5b, 0xeb);
+
+DECLARE_TR_CTX(ams_tr, SOF_UUID(ams_uuid), LOG_LEVEL_INFO);
+
+/* AMS routing table */
+const uint8_t g_ams_core_routing[CONFIG_CORE_COUNT][CONFIG_CORE_COUNT] = {
+/* to core: 0..N */
+/* from core 0 */ {
+	0,
+	1,
+	2,
+	}, /* primary core can target every other core directly */
+/* from core 1 */ {
+	0,
+	1,
+	0,
+	}, /* and secondary must always go through primary core */
+/* from core 2 */ {
+	0,
+	0,
+	2,
+}};
```

Comment:
Hmm, this is kind of assumes CONFIG_CORE_COUNT is 3 , right? Either you'd get build warning/errors or at minimum, not the expected functionality.

Wouldn't it be easier to expose this fairly simple logic (everything needs to go via core0) without tables?